File: compiled_lang/java/src/o20.longform.java
import java.io.PrintStream;
import java.util.ArrayList;
import java.util.List;

// Not public - see a00.output.java for why.
class O20Longform {
    private static void usage(PrintStream out, String scriptName) {
        out.println();
        out.println("Usage: " + scriptName + " [--coffee|-c N] [--espresso|-e N] [--latte|-l N] [--macchiato|-k N] [--capucino|-p N] [--mocha|-m N] [--tea|-t N] [--help|-h|-?]");
        out.println();
        out.println("  --coffee,    -c N  Coffee");
        out.println("  --espresso,  -e N  Espresso");
        out.println("  --latte,     -l N  Latte");
        out.println("  --macchiato, -k N  Machiato");
        out.println("  --capucino,  -p N  Capucino");
        out.println("  --mocha,     -m N  Mocha");
        out.println("  --tea,       -t N  Tea");
        out.println("  --help,      -h    Display this help message");
        out.println("  -?                 Display this help message");
        out.println();
    }

    public static void main(String[] args) {
        String scriptName = System.getProperty("invoked.as");

        if (args.length == 0) {
            usage(System.err, scriptName);
            System.exit(1);
            return;
        }

        List<String> names = new ArrayList<>();
        List<String> counts = new ArrayList<>();

        int i = 0;
        while (i < args.length) {
            switch (args[i]) {
                case "--coffee": case "-c":
                    names.add("coffee"); counts.add(args[i + 1]); i += 2; break;
                case "--espresso": case "-e":
                    names.add("espresso"); counts.add(args[i + 1]); i += 2; break;
                case "--latte": case "-l":
                    names.add("latte"); counts.add(args[i + 1]); i += 2; break;
                case "--macchiato": case "-k":
                    names.add("macchiato"); counts.add(args[i + 1]); i += 2; break;
                case "--capucino": case "-p":
                    names.add("capucino"); counts.add(args[i + 1]); i += 2; break;
                case "--mocha": case "-m":
                    names.add("mocha"); counts.add(args[i + 1]); i += 2; break;
                case "--tea": case "-t":
                    names.add("tea"); counts.add(args[i + 1]); i += 2; break;
                case "--help": case "-h": case "-?":
                    usage(System.out, scriptName);
                    return;
                default:
                    usage(System.err, scriptName);
                    System.exit(1);
                    return;
            }
        }

        if (names.isEmpty()) {
            usage(System.err, scriptName);
            System.exit(1);
            return;
        }

        System.out.println();
        System.out.println("You ordered: ");
        for (int idx = 0; idx < names.size(); idx++) {
            int n = Integer.parseInt(counts.get(idx));
            String label = names.get(idx);
            if (n != 1) label += "s";
            System.out.println("* " + counts.get(idx) + " " + label);
        }
    }
}

File: lessons/win_scripts/wsh.jscript/scripts/o20.longform.js
var usage = "\n" +
            "Usage: " + WScript.ScriptName + " [--coffee|-c N] [--espresso|-e N] [--latte|-l N] [--macchiato|-k N] [--capucino|-p N] [--mocha|-m N] [--tea|-t N] [--help|-h|-?]\n" +
            "\n" +
            "  --coffee,    -c N  Coffee\n" +
            "  --espresso,  -e N  Espresso\n" +
            "  --latte,     -l N  Latte\n" +
            "  --macchiato, -k N  Machiato\n" +
            "  --capucino,  -p N  Capucino\n" +
            "  --mocha,     -m N  Mocha\n" +
            "  --tea,       -t N  Tea\n" +
            "  --help,      -h    Display this help message\n" +
            "  -?                 Display this help message\n" +
            "\n";

var flags = {
   "--coffee": "coffee", "-c": "coffee",
   "--espresso": "espresso", "-e": "espresso",
   "--latte": "latte", "-l": "latte",
   "--macchiato": "macchiato", "-k": "macchiato",
   "--capucino": "capucino", "-p": "capucino",
   "--mocha": "mocha", "-m": "mocha",
   "--tea": "tea", "-t": "tea"
};

var orders = [];
var i = 0;
while (i < WScript.Arguments.length) {
   var arg = WScript.Arguments(i);
   if (arg == "-h" || arg == "-?" || arg == "--help") {
      WScript.stdout.write(usage);
      WScript.Quit(0);
   } else if (flags.hasOwnProperty(arg)) {
      var name = flags[arg];
      var n = parseInt(WScript.Arguments(i + 1), 10);
      var suffix = (n == 1) ? "" : "s";
      orders.push(n + " " + name + suffix);
      i += 2;
   } else {
      WScript.stderr.write(usage);
      WScript.Quit(1);
   }
}

if (orders.length == 0) {
   WScript.stderr.write(usage);
   WScript.Quit(1);
}

WScript.Echo("");
WScript.Echo("You ordered: ");
for (var j = 0; j < orders.length; j++) {
   WScript.Echo("* " + orders[j]);
}
